### ISSUE TITLE
bfgminer ignores command line settings

### ISSUE THREAD

**Issue by appleguru** [luke-jr/bfgminer#387]
if there exists a bfgminer or cgminer config file, bfgminer will read it and override command line arguments provided at runtime... very annoying.

Also, would be great to be able to --set-device antminer:freq=HEX per-device, as some devices reliably clock higher than others... does the ant miner return any serial number/unique identifier that could be used to track per-device settings?

**Comment by luke-jr** [comment]
Currently, BFGMiner loads the default config file _after_ parsing the command line, _only if_ the user did not specify a config file of his own. I'm willing to consider different behaviour, but please suggest and argue why it should change. Keep in mind that we cannot have the "only load default config if another is not specified" rule _and also_ load the default _before_ the command line is parsed.

**Comment by appleguru** [comment]
> loads the default config file after parsing the command line, **only if the user did not specify a config file of his own**

Expected behavior is that if I don't specify a config file, you use the command line arguments provided (and do NOT load any config file at all, default or otherwise). The only time you should be loading the default config file is if NO command line arguments are provided (IMO).

What I am seeing with 3.10.0 on OS X is that if there is a config file at ~/.bfgminer/bfgminer.conf or at ~/.cgminer/cgminer.conf, bfgminer will ALWAYS load this file, overwriting whatever command line options are specified at runtime. Only way to avoid this is to remove all default configs or specify a different config file with --config or -c at runtime.

If you MUST load the default config every time --config is omitted, the configuration parameters provided on the command line by the user should still take precedent. Ideally the two configuration sets would be merged (with any duplicate/conflicting parameters coming from the command line)

**Comment by luke-jr** [comment]
How about we just add a --no-config option?

**Comment by appleguru** [comment]
That'd work for me, though may not solve the frustrations of new users that are expecting one thing and getting another... IMO, the default behavior should be --no-config if none is specified.

I guess I just don't really see an upside to ever overriding what a user spent time typing in with some defaults loaded from a mysterious (often hidden) location... seems very backwards to me, especially when those defaults can come from multiple places / another program... (cgminer).

**Comment by luke-jr** [comment]
Not loading it with options, makes debugging harder (output from -d? -D would differ from reality).
Loading it before parsing options, means we cannot disable it at all (ie, --no-config and a different --config wouldn't disable the default anymore).

So, I'm not sure there is a perfect solution :(

**Comment by appleguru** [comment]
Perhaps in addition to a --no-config option, add line in the curses window indicating where the active config file (if any) was loaded from? That would at least help clear up the confusion on where settings are coming from.

Something like "Configuration loaded from: /path/to/bfgminer.conf"

**Comment by luke-jr** [comment]
If I can update ccan, I found a way to parse the default config first but still disable it if there's a config (or --no-config) specified. However, I am also having this prioritise defaultconfig pools _last_...

I've pushed the _default_config_first_ branch to GitHub if you'd like to test it out. However, upstream ccan is broken for Windows right now, so until that is fixed, I cannot merge this. :(

(I did write patches to fix it and sent them to the author of ccan, so hopefully it will be in time for the next release).
